MobileWallet/Backup/ICloudBackup.swift

```diff
@@ -0,0 +1,361 @@
+                    completion(ICloudBackupWalletError.privateKeyError)
+                    return
+                }
+                completion(nil)
+            }
+        } catch {
+            completion(error)
+        }
+    }
+}
+
+// MARK: - private methods
+extension ICloudBackup {
+    private func restoreBackup(walletFolder: String, to directory: URL, completion: @escaping ((_ error: Error?) -> Void)) {
+        downloadBackup(walletFolder: walletFolder) { [weak self] (url, error) in
+            guard let zippedBackupURL = url, error == nil else {
+                completion(error)
+                return
+            }
+
+            do {
+                try self?.unzipBackup(url: zippedBackupURL, to: directory)
+                completion(nil)
+            } catch {
+                completion(error)
+            }
+        }
+    }
+
+    private func downloadBackup(walletFolder: String, completion: @escaping ((_ url: URL?, _ error: Error?) -> Void)) {
+        let iCloudFolderURL: URL!
+        var urls = [URL]()
+
+        do {
+            iCloudFolderURL = try iCloudDirectory().appendingPathComponent(walletFolder)
+            urls = try FileManager.default.contentsOfDirectory(atURL: iCloudFolderURL, sortedBy: .created, options: [])
+        } catch {
+            completion(nil, error)
+        }
+
+        if let backupUrl = urls.last {
+            var lastPathComponent = backupUrl.lastPathComponent
+            let folderPath = backupUrl.deletingLastPathComponent().path
+            // if the last path component contains the “.icloud” extension. If yes the file is not on the device else the file is already downloaded.
+
+            if lastPathComponent.contains(".icloud") {
+                lastPathComponent.removeFirst()
+                let downloadedFilePath = folderPath + "/" + lastPathComponent.replacingOccurrences(of: ".icloud", with: "")
+
+                do {
+                    try FileManager.default.startDownloadingUbiquitousItem(at: backupUrl)
+                } catch {
+                    completion(nil, error)
+                }
+
+                Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { (timer) in
+                    if FileManager.default.fileExists(atPath: downloadedFilePath) {
+                        timer.invalidate()
+                        completion(URL(fileURLWithPath: downloadedFilePath), nil)
+                    }
+                }
+            } else {
+                completion(backupUrl, nil)
+            }
+        }
+    }
+
+    private func getLastWalletBackupFolder() throws -> String {
+        var wallets = [String]()
+        let iCloudFolderURL = try iCloudDirectory()
+        do {
+            let urls = try FileManager.default.contentsOfDirectory(atURL: iCloudFolderURL, sortedBy: .created, options: [])
+            urls.forEach({
+                if $0.lastPathComponent != "Documents" {
```

Comment:
I think this is satisfactory, but a regex check would make it much more solid.

Reply:
i think, regex will be superfluous. Honestly this check is superfluous, i have changed directory of saving backup. So i'll just delete this fragment of code

Reply:
Ok.